backend/corpora/dataset_submissions/app.py

```diff
@@ -6,14 +6,18 @@
 
 from sqlalchemy.orm import Session
 
-from backend.corpora.common.entities import Collection
+from backend.corpora.common.entities import Dataset
 from backend.corpora.common.upload import upload
 from backend.corpora.common.utils.db_session import db_session_manager
 from backend.corpora.common.utils.exceptions import CorporaException
 
 logger = logging.getLogger(__name__)
-USERNAME_REGEX = r"[\w\-\|]+"
+USERNAME_REGEX = r"(?P<username>[\w\-\|]+)"
 UUID_REGEX = r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+EXTENSION_REGEX = r"(?P<extension>h5ad)"
+DATASET_ID_REGEX = f"(?P<dataset_uuid>{UUID_REGEX})"
+COLLECTION_ID_REGEX = f"(?P<collection_uuid>{UUID_REGEX})"
+REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((?P<tag>.*)|{DATASET_ID_REGEX}).{EXTENSION_REGEX}$"
```

Comment:
gorgeous 🤩

Reply:
But one problem. suggested fix:

```suggestion
REGEX = f"^{USERNAME_REGEX}/{COLLECTION_ID_REGEX}/((?P<tag>.*)|{DATASET_ID_REGEX})\\.{EXTENSION_REGEX}$"
```

Reply:
I thought I fixed that. Good catch.